### PR TITLE
Return Never in Option "or" closure

### PR DESCRIPTION
**Pull request by DimaZava**
### Added

- **Optional**
  - Added `unwrapped(or error: Error)` to get the wrapped value of an optional. If the optional is `nil`, calls Never-returning closure.

**Note**: I didn't add tests since all the assertions shouldn't be by default in production code. Also I know only one tough way to implement global default Never returning functions replacement in order to organise proper testing of ones. The other option is to use 3rd party library with similar testing abilities. Feel free to comment and propose more suitable solutions if possible.

## Checklist

- [x] I checked the [**Contributing Guidelines**](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md) before creating this request.
- [x] New extensions are written in Swift 5.0.
- [x] New extensions support iOS 8.0+ / tvOS 9.0+ / macOS 10.10+ / watchOS 2.0+, or use `@available` if not.
- [ ] I have added tests for new extensions, and they passed.
- [x] All extensions have a **clear** comments explaining their functionality, all parameters and return type in English.
- [x] All extensions are declared as **public**.
- [x] I have added a [changelog](https://github.com/SwifterSwift/SwifterSwift/blob/master/CHANGELOG_GUIDELINES.md) entry describing my changes.

### CODE REVIEW

File: Sources/SwifterSwift/SwiftStdlib/OptionalExtensions.swift [hunk 1]
```diff
@@ -40,6 +40,21 @@ public extension Optional {
         return wrapped
     }
 
+    /// SwifterSwift: Gets the wrapped value of an optional. If the optional is `nil`, calls Never-returning closure.
+    ///
+    ///        let foo: String? = nil
+    ///        foo.or { preconditionFailute() } -> Throws preconditionFailure with default description
```
Comment (kwstasna) [inline Sources/SwifterSwift/SwiftStdlib/OptionalExtensions.swift]: change this `preconditionFailute()` to `preconditionFailure()`. Just a small typo.

**Comment by guykogus** [comment]
@DimaZava can you please try add some tests?

**Comment by DimaZava** [comment]
@guykogus I didn't add tests since all the assertions shouldn't be by default in production code. Also I know only one tough way to implement global default Never returning functions replacement in order to organise proper testing of ones. The other option is to use 3rd party library with similar testing abilities. Feel free to comment and propose more suitable solutions if possible.

File: Sources/SwifterSwift/SwiftStdlib/OptionalExtensions.swift [hunk 2]
```diff
@@ -40,6 +40,21 @@ public extension Optional {
         return wrapped
     }
 
+    /// SwifterSwift: Gets the wrapped value of an optional. If the optional is `nil`, calls Never-returning closure.
+    ///
+    ///        let foo: String? = nil
+    ///        foo.or { preconditionFailure() } -> Throws preconditionFailure with default description
+    ///
+    ///        let bar: String? = "bar"
+    ///        bar.or { fatalError("Error description") } -> Throws fatal error with custom description
```
[the remark this answers is not in the document]
Comment (DimaZava) [reply Sources/SwifterSwift/SwiftStdlib/OptionalExtensions.swift]: @LucianoPAlmeida I've attempted to develop an idea of using such shorthand operators for Optional's. This way we can discuss `func unwrapped(or defaultValue: Wrapped) -> Wrapped` necessity, which actually increases both code size and complexity comparing to `??` operator. On the other hand I support your position, perhaps some operators, use cases, etc. like one I provided should be implemented only in explicit way to be obvious for one who works with such "fatal" code.

[the remark this answers is not in the document]
Comment (omaralbeik) [reply Sources/SwifterSwift/SwiftStdlib/OptionalExtensions.swift]: @LucianoPAlmeida personally I'm ok with adding this, in the end, it is one way of handling errors in optionals, and it does not rely on other extensions

[the remark this answers is not in the document]
Comment (LucianoPAlmeida) [reply Sources/SwifterSwift/SwiftStdlib/OptionalExtensions.swift]: Sure @omaralbeik!

**Comment by LucianoPAlmeida** [comment]
I'm still a bit skeptical about this one, for both the reasons I've mentioned above and the fact I can't imagine anyone using it really... I can't think in a use case or reason for that. 
I mean, it doesn't hurt adding but also I can't see the benefit too.

**Comment by guykogus** [comment]
If anything I think that this extension makes error handling worse. When you're calling a `Never` function it's because you've hit a dead end and want to exit the function (or the entire program, probably). If you think of the examples where this generally happens with optionals, you have something like:
```Swift
func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    guard let cell = tableView.dequeueReusableCell(withIdentifier: "identifier", for: indexPath) as? MyTableViewCell else {
        preconditionFailure()
    }
    cell.foo()
    return cell
}
```

In this function the return value is non-optional, so if you use this extension you still need to handle the unwrapping.
```Swift
func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = tableView.dequeueReusableCell(withIdentifier: "identifier", for: indexPath) as? MyTableViewCell
    cell.or { preconditionFailure() }
    cell!.foo()
    return cell!
}
```

The built-in language features for handling this scenario are far better. I can't see a scenario in which the new extension would be preferred. @DimaZava can you suggest any scenario where the extension is preferred?

**Comment by LucianoPAlmeida** [comment]
I think its safe to say that this one could be closed :) But @DimaZava if you still have some points to discuss feel free to re-open it 👍

**Pull request closed without merging**